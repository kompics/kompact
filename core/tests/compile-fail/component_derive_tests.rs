use kompact::prelude::*;

#[derive(Debug)]
struct TestPort;
impl Port for TestPort {
    type Indication = String;
    type Request = usize;
}

#[derive(ComponentDefinition)] //~ ERROR: Ambiguous port type: There are multiple fields with type TestPort (["provided_test_port1", "provided_test_port2"]). You cannot derive ComponentDefinition in these cases, as you must resolve the ambiguity manually.
struct DuplicatePortComponent {
    ctx: ComponentContext<Self>,
    provided_test_port1: ProvidedPort<TestPort>,
    provided_test_port2: ProvidedPort<TestPort>,
    required_test_port1: RequiredPort<TestPort>,
    required_test_port2: RequiredPort<TestPort>,
}

fn main() { 
    // unused 
}
