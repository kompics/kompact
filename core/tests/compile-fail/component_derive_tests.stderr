error: Ambiguous port type: There are multiple fields with type TestPort (["provided_test_port1", "provided_test_port2"]). You cannot derive ComponentDefinition in these cases, as you must resolve the ambiguity manually.
  --> tests/compile-fail/component_derive_tests.rs:10:10
   |
10 | #[derive(ComponentDefinition)] //~ ERROR: Ambiguous port type: There are multiple fields with type TestPort (["provided_test_port1", "pro...
   |          ^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `ComponentDefinition` (in Nightly builds, run with -Z macro-backtrace for more info)

error: Ambiguous port type: There are multiple fields with type TestPort (["required_test_port1", "required_test_port2"]). You cannot derive ComponentDefinition in these cases, as you must resolve the ambiguity manually.
  --> tests/compile-fail/component_derive_tests.rs:10:10
   |
10 | #[derive(ComponentDefinition)] //~ ERROR: Ambiguous port type: There are multiple fields with type TestPort (["provided_test_port1", "pro...
   |          ^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `ComponentDefinition` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0407]: method `provided_ref` is not a member of trait `RequireRef`
  --> tests/compile-fail/component_derive_tests.rs:10:10
   |
10 | #[derive(ComponentDefinition)] //~ ERROR: Ambiguous port type: There are multiple fields with type TestPort (["provided_test_port1", "pro...
   |          ^^^^^^^^^^^^^^^^^^^ not a member of trait `RequireRef`
   |
   = note: this error originates in the derive macro `ComponentDefinition` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0407]: method `connect_to_required` is not a member of trait `RequireRef`
  --> tests/compile-fail/component_derive_tests.rs:10:10
   |
10 | #[derive(ComponentDefinition)] //~ ERROR: Ambiguous port type: There are multiple fields with type TestPort (["provided_test_port1", "pro...
   |          ^^^^^^^^^^^^^^^^^^^
   |          |
   |          not a member of trait `RequireRef`
   |          help: there is an associated function with a similar name: `connect_to_provided`
   |
   = note: this error originates in the derive macro `ComponentDefinition` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `DuplicatePortComponent: NetworkActor` is not satisfied
  --> tests/compile-fail/component_derive_tests.rs:12:10
   |
12 |     ctx: ComponentContext<Self>,
   |          ^^^^^^^^^^^^^^^^^^^^^^ the trait `NetworkActor` is not implemented for `DuplicatePortComponent`
   |
   = note: required for `DuplicatePortComponent` to implement `kompact::prelude::Actor`
   = note: required for `DuplicatePortComponent` to implement `ActorRaw`
   = note: required for `DuplicatePortComponent` to implement `ComponentTraits`
note: required by a bound in `kompact::component::ComponentContext`
  --> src/component/context.rs
   |
   | pub struct ComponentContext<CD: ComponentTraits> {
   |                                 ^^^^^^^^^^^^^^^ required by this bound in `kompact::component::ComponentContext`

error[E0277]: the trait bound `DuplicatePortComponent: ComponentLifecycle` is not satisfied
  --> tests/compile-fail/component_derive_tests.rs:10:10
   |
10 | #[derive(ComponentDefinition)] //~ ERROR: Ambiguous port type: There are multiple fields with type TestPort (["provided_test_port1", "pro...
   |          ^^^^^^^^^^^^^^^^^^^ the trait `ComponentLifecycle` is not implemented for `DuplicatePortComponent`
   |
   = help: the following other types implement trait `ComponentLifecycle`:
             BigPingerAct
             BigPongerAct
             DeadletterBox
             ForwarderAct
             LocalDispatcher
             NetworkDispatcher
             NetworkStatusCounter
             PingStream
           and 4 others
note: required by a bound in `kompact::component::ComponentDefinition`
  --> src/component/definition.rs
   |
   | pub trait ComponentDefinition: DynamicComponentDefinition
   |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `kompact::component::ComponentDefinition`
   = note: this error originates in the derive macro `ComponentDefinition` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `DuplicatePortComponent: NetworkActor` is not satisfied
  --> tests/compile-fail/component_derive_tests.rs:10:10
   |
10 | #[derive(ComponentDefinition)] //~ ERROR: Ambiguous port type: There are multiple fields with type TestPort (["provided_test_port1", "pro...
   |          ^^^^^^^^^^^^^^^^^^^ the trait `NetworkActor` is not implemented for `DuplicatePortComponent`
   |
   = help: the following other types implement trait `ActorRaw`:
             TestComponent1
             TestComponent2
   = note: required for `DuplicatePortComponent` to implement `kompact::prelude::Actor`
   = note: required for `DuplicatePortComponent` to implement `ActorRaw`
note: required by a bound in `kompact::component::ComponentDefinition`
  --> src/component/definition.rs
   |
   | pub trait ComponentDefinition: DynamicComponentDefinition
   |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `kompact::component::ComponentDefinition`
   = note: this error originates in the derive macro `ComponentDefinition` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `DuplicatePortComponent: NetworkActor` is not satisfied
  --> tests/compile-fail/component_derive_tests.rs:10:10
   |
10 | #[derive(ComponentDefinition)] //~ ERROR: Ambiguous port type: There are multiple fields with type TestPort (["provided_test_port1", "pro...
   |          ^^^^^^^^^^^^^^^^^^^ the trait `NetworkActor` is not implemented for `DuplicatePortComponent`
   |
   = note: required for `DuplicatePortComponent` to implement `kompact::prelude::Actor`
   = note: required for `DuplicatePortComponent` to implement `ActorRaw`
   = note: required for `DuplicatePortComponent` to implement `ComponentTraits`
note: required by a bound in `kompact::component::Component`
  --> src/component/actual_component.rs
   |
   | pub struct Component<CD: ComponentTraits> {
   |                          ^^^^^^^^^^^^^^^ required by this bound in `kompact::component::Component`
   = note: this error originates in the derive macro `ComponentDefinition` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `DuplicatePortComponent: NetworkActor` is not satisfied
  --> tests/compile-fail/component_derive_tests.rs:10:10
   |
10 | #[derive(ComponentDefinition)] //~ ERROR: Ambiguous port type: There are multiple fields with type TestPort (["provided_test_port1", "pro...
   |          ^^^^^^^^^^^^^^^^^^^ the trait `NetworkActor` is not implemented for `DuplicatePortComponent`
   |
   = note: required for `DuplicatePortComponent` to implement `kompact::prelude::Actor`
   = note: required for `DuplicatePortComponent` to implement `ActorRaw`
   = note: required for `DuplicatePortComponent` to implement `ComponentTraits`
note: required by a bound in `kompact::component::ComponentContext`
  --> src/component/context.rs
   |
   | pub struct ComponentContext<CD: ComponentTraits> {
   |                                 ^^^^^^^^^^^^^^^ required by this bound in `kompact::component::ComponentContext`
   = note: this error originates in the derive macro `ComponentDefinition` (in Nightly builds, run with -Z macro-backtrace for more info)
